Changes to restore multiprocessing and unit tests

mwregan2 opened the pull request:
This PR restores the multiprocessing capability to the jump and adds back the unit tests for the twopoint_difference.py. Note that there is a parallel PR for the JWST repo that is required for this to work.

kmacdonald-stsci reviewed (commented):
Besides style failures it looks good.

nden commented:
@mwregan2 I see the tests in this package are in stcal/tests, the top directory and there're tests for two point differences. I did not check if they are the same tests though.

Also did you compare results from processing a data set with multiprocessing on and off. It was previously disabled because it gave different results.

hbushouse commented on src/stcal/jump/jump.py:
```diff
@@ -129,33 +129,45 @@ def detect_jumps(frames_per_group, data, gdq, pdq, err,
-    n_slices = 1
-
-    yinc = int(n_rows / n_slices)
+    print('new stcal jump')
```
can't have print statements in production code. Either remove or replace with log messages.

mwregan2 commented:
The unit tests are different from what is in the JWST repo. These test the twopoint_difference routine.

mwregan2 commented:
Nadia,
  Are there tests that show the differences? That would really help in finding the problem. I’ll add some tests to see if I can find any differences.
Mike

From: Nadia Dencheva ***@***.***>
Reply-To: spacetelescope/stcal ***@***.***>
Date: Monday, May 2, 2022 at 10:13 AM
To: spacetelescope/stcal ***@***.***>
Cc: Michael Regan ***@***.***>, Mention ***@***.***>
Subject: Re: [spacetelescope/stcal] Changes to restore multiprocessing and unit tests (PR #85)


@mwregan2<https://urldefense.com/v3/__https:/github.com/mwregan2__;!!CrWY41Z8OgsX0i-WU-0LuAcUu2o!w-8rwDIUBDUhrQj6B020MyNPewMTKxPOgD15X_qBOAWpcFFeNzA2OdL9PvO04ucuNSUgmdtrQWtbKAvsAup9y_4$> I see the tests in this package are in stcal/tests, the top directory and there're tests for two point differences. I did not check if they are the same tests though.

Also did you compare results from processing a data set with multiprocessing on and off. It was previously disabled because it gave different results.

—
Reply to this email directly, view it on GitHub<https://urldefense.com/v3/__https:/github.com/spacetelescope/stcal/pull/85*issuecomment-1114934138__;Iw!!CrWY41Z8OgsX0i-WU-0LuAcUu2o!w-8rwDIUBDUhrQj6B020MyNPewMTKxPOgD15X_qBOAWpcFFeNzA2OdL9PvO04ucuNSUgmdtrQWtbKAvs_tmvjuQ$>, or unsubscribe<https://urldefense.com/v3/__https:/github.com/notifications/unsubscribe-auth/AF6KJLVHETQJWPER2EY33YTVH7PIZANCNFSM5U2OOMOA__;!!CrWY41Z8OgsX0i-WU-0LuAcUu2o!w-8rwDIUBDUhrQj6B020MyNPewMTKxPOgD15X_qBOAWpcFFeNzA2OdL9PvO04ucuNSUgmdtrQWtbKAvsuRUhCzw$>.
You are receiving this because you were mentioned.Message ID: ***@***.***>

nden commented:
@mwregan2 Sorry for the late reply. I see you added a test in jwst comparing single vs multiprocessing run and it passes. 

[test_single_CR_neighbor_flag](https://github.com/spacetelescope/jwst/blob/4316c729f5c0f3c1734feac1a4ec9dad327498a5/jwst/jump/tests/test_detect_jumps.py#L737) gives different results if run in multiprocessing model. I don't know the details of the algorithm to judge why one is passing and the other fails.

mwregan2 commented:
I fixed the issue with multiprocessing giving different results. This was caused by the code changing the input model and then the same model being sent back in for the second call. This could not happen in the operational system.
  Even so, I fixed the problem and added a test case that fails without the fix. I just needed to have a new variable for the updated gdq array.

nden commented:
Superseded by #87
